Add contrib ports (Contrib ports part 2)

ypujante opened the pull request:
This is part 2 of contrib ports as we discussed.

In this part I did the following:

* add support for ports in tools/port/contrib folder
* added a README.md in this folder describing the requirements for contrib ports
* added a test in test_other to make sure that contrib port are read properly (this test uses the only port available at the moment, and can be changed once another port is available)
* updated the documentation + added a new Contrib-Ports.rst documentation page which is automatically generated via a new script (since it is not a settings, I could not put in under the settings generated page). Note that IMO I think it would make a lot of sense to have a Ports.rst documentation instead which contain both "official" ports and contrib ports, but I wasn't sure if that was a direction you would want to go as this will require a lot more changes to either settings.js or the official ports (even if the changes are essentially documentation...)
* changed CI to run check on this new documentation
* had to change `link.py` so that my port would compile as it uses a js library and the order in which things were happening was wrong

Finally:

* I did NOT load the port on demand, so contrib ports are loaded at boot time. Changing to load contrib ports on demand would be easy but it is more changes and I know you were not too keen on this at this moment. Could be done separately and/or later if there is a need.
* I did NOT rename my port (you had mentioned this previously) because with this new setup/option its name is `contrib.glfw3` which makes it pretty obvious that it is not the build-in one

ypujante commented on tools/ports/__init__.py:
```diff
@@ -34,13 +34,11 @@
 def load_port(name):
-  expected_attrs = ['get', 'clear', 'show']
-  port = __import__(name, globals(), level=1)
+  port = __import__(name, globals(), level=1, fromlist=[None])
```
I will double check if I can get away without it, but I copied this from your own PR

ypujante replied:
So I double checked and without ` fromlist=[None]` it doesn't work. Like I said I had copied this bit from your PR and I assumed you had done it for a good reason. Without it, the modules under contrib are not loaded properly. 

Displaying `port.__name__` which is the name of the module, shows that for modules under ports it works (I get for example `tools.ports.sdl2`) but for ports under `contribs` it doesn't as the name is `tools.ports.contrib`.

If I use ` fromlist=[None]`, then the name is now correct `tools.ports.contrib.glfw3`

Asking Jetbrains AI give this answer (which I clipped because it is quite long but quite informative):

> In conclusion, fromlist=[None] is a way to tell `__import__` to treat the last part of the module path as the actual module that we want to import. If the list is empty or not provided `__import__` will instead return the first module in the provided path.

Note that documentation for `__import__` recommends: _Because this function is meant for use by the Python interpreter and not for general use, it is better to use `importlib.import_module()` to programmatically import a module._

ypujante commented on tools/ports/__init__.py:
```diff
@@ -57,13 +55,32 @@ def load_port(name):
   if not hasattr(port, 'variants'):
     # port variants (default: no variants)
     port.variants = {}
+  if not hasattr(port, 'show') and port.is_contrib:
```
the auto generated `show` method relies on `port.project_license()` (or `port.license()`) which is required only for contrib ports. That being said if `port.license()` is not implemented then CI sanity checks will fail when it invokes `show_ports()` so it's safe to remove this check anyway because it would be caught

ypujante commented on tools/ports/__init__.py:
```diff
@@ -57,13 +55,32 @@ def load_port(name):
+def validate_port(port):
+  expected_attrs = ['get', 'clear', 'show']
+  if port.is_contrib:
+    expected_attrs += ['project_url', 'project_description', 'project_license']
```
that's fine with me

ypujante commented on site/source/docs/compiling/Building-Projects.rst:
```diff
@@ -241,15 +241,18 @@ To see a list of all available ports, run ``emcc --show-ports``.
+Contrib ports
+-------------
+
+Contrib ports are contributed by the wider community and supported on a "best effort" basis. Since they are not run as part of emscripten CI they are not always guaranteed to build or function. See :ref:`Contrib Ports <contrib_ports>` for more information.
```
sure

sbc100 commented:
> @sbc100
> 
> Looking at the test failures:
> 
> * do you know why `build-docs` fail? The error is this:
> 
> ```
> tools/maint/update_settings_docs.py --check
> ....
> tools/maint/update_contrib_ports_docs.py --check
> update_contrib_ports_docs: warning: config file not found: /root/project/.emscripten.  You can create one by hand or run `emcc --generate-config`
> update_contrib_ports_docs: error: LLVM_ROOT not set in config (/root/project/.emscripten), and `clang` not found in PATH
> 
> Exited with code exit status 1
> ```
> 
> Running `update_settings_docs.py --check` works but not the new one `tools/maint/update_contrib_ports_docs.py --check`. Did I miss something?
> 
> Then there is this one in test-browser-chrome-wasm64:
> 
> ```
> ======================================================================
> ERROR [61.376s]: test_offset_converter (test_browser.browser64)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/root/project/test/common.py", line 1989, in run_browser
>     self.assertContained(expected, output)
>   File "/root/project/test/common.py", line 1321, in assertContained
>     additional_info
>   File "/usr/lib/python3.6/unittest/case.py", line 670, in fail
>     raise self.failureException(msg)
> AssertionError: Expected to find '/report_result?exit:1
> ' in '[no http server activity]
> ', diff:
> 
> --- expected
> +++ actual
> @@ -1 +1 @@
> -/report_result?exit:1
> +[no http server activity]
> ```
> 
> I will try to run locally. I hope it is not due to the change I made in `link.py`

How about we just land the initial version of this change without the auto-updating-docs thing.  Its a fair amount of complexity that I'm not sure is worth it.  We can then discussion how/if to document the contrib ports as a followup ?

sbc100 commented:
The failure here is because you are importing `shared.py` which is what triggers the check to see of emcc is setup correctly.

Ideally these checks would not be done at import time, but that is the way it is, mostly for historical reasons.

Note that `tools/maint/update_settings_docs.py` does not import `shared.py`.

ypujante commented:
> How about we just land the initial version of this change without the auto-updating-docs thing. Its a fair amount of complexity that I'm not sure is worth it. We can then discussion how/if to document the contrib ports as a followup ?

I suppose that is an "easy" fix. Personally I feel like setting up good hygiene right from the start is a good thing. The idea of contrib ports is that there is very little burden when one is added for you/whoever else review the changes. If the documentation is automatically updated, that is one less thing to worry about. But I get your point if you prefer for me to simplify it for now/then do another PR later.

sbc100 commented:
> > How about we just land the initial version of this change without the auto-updating-docs thing. Its a fair amount of complexity that I'm not sure is worth it. We can then discussion how/if to document the contrib ports as a followup ?
> 
> I suppose that is an "easy" fix. Personally I feel like setting up good hygiene right from the start is a good thing. The idea of contrib ports is that there is very little burden when one is added for you/whoever else review the changes. If the documentation is automatically updated, that is one less thing to worry about. But I get your point if you prefer for me to simplify it for now/then do another PR later.

I see your point.   I'm was think its would be good to find a way for you to get the parts that you can land in quickly without too much back and forth or adding more to a single PR.

I'm also not too worried about whether contrib ports get documented, at least no initially.   Since we are not supporting these ports, there could even be an argument that we should not put them on the website at all(?)

ypujante commented:
> I'm also not too worried about whether contrib ports get documented, at least no initially. Since we are not supporting these ports, there could even be an argument that we should not put them on the website at all(?)

I would argue that contrib ports are value add to the emscripten platform with little hassle to the project itself, so I would look at it more as a win: each new contrib port makes emscripten better to use. Obviously I am not talking about the documentation about the contrib port itself, but the basic information like port name/short description/url/license (which is what I generated) in an easy to browse list seems to me very valuable. \

As a user of the emscripten platform, if I want to know if there is a port of XXX, then I can just look at this single page (and since the documentation as search built-in it is easier to find). Otherwise how will I know that there is a port of XXX that is super easy to use (--use-port) option? Sure you can run `emcc --show-ports` but that means installing emscripten to get this list... 

For a concrete example, when you added libarchive to your PR about this topic a while ago, it would be great, when it is added that it just appears in the list. If I needed to use zip compression in my code it would be very valuable for me to just check/search the documentation to see if there is port that I can simply use even if not supported officially.

ypujante commented:
I have addressed the code review and removed auto generation of contrib ports documentation. I think that it still should be done since it is kind of easy, the only catch is that I could not figure out how to run the script due to the `.emscripten` missing file and I couldn't see a way to quickly address this problem...

ypujante commented:
All tests are passing. If you are satisfied with the changes, I can add an entry to the changelog, something like:

> Added concept of contrib ports which are ports contributed by the wider community and supported on a "best effort" basis. A first contrib port is available via `--use-port=contrib.glfw3`: an emscripten port of glfw written in C++ with many features like support for multiple windows. (#21244)

sbc100 commented on tools/ports/contrib/glfw3.py:
```diff
@@ -0,0 +1,61 @@
+HASH = 'c3c96718e5d2b37df434a46c4a93ddfd9a768330d33f0d6ce2d08c139752894c2421cdd0fefb800fe41fafc2bbe58c8f22b8aa2849dc4fc6dde686037215cfad'
+
+# name is set when the port is read
+name = ''
```
I don't think you need these two lines do you?

ypujante replied:
It is technically not required, but I use it in the rest of this code and the IDE is not happy about it because it doesn't see it. It is also a reminder that something is setting it which I think is not obvious (for people who wants to add a contrib port) when you look at the code.

sbc100 replied:
I think its more confusing though.. since setting a name here would be more of an error.    The name is set based on the filename alone. 

I'm not sure you should depend on and reference `name` within this file... at least its not something we do any of the existing ports.

ypujante replied:
I renamed the variable to avoid confusion. I don't like to repeat myself like the other ports do (for example in `bullet.py`, the name `bullet` is used many times in places where it has to be the same and all the other ports follow this "bad" pattern IMO) and the value needs to be unique across ports otherwise there is the potential of stepping on each other feet, which is why I was using `name` because it is unique across ports.

But since that causes confusion I renamed it.

sbc100 commented on tools/ports/contrib/glfw3.py:
```diff
@@ -0,0 +1,61 @@
+
+
+def license():
+  return 'Apache 2.0 license'
```
Should we make these last 3 into globals rather than functions?   (and maybe put them up the top of the file along with TAG and HASH?

ypujante replied:
sounds good

ypujante replied:
It looks cleaner/more concise and right at the top when opening the file. Definitely better.

sbc100 reviewed (commented):
lgtm with a couple of last minor comments.

Please go ahead and add that changelog entry.

sbc100 commented on tools/ports/__init__.py:
```diff
@@ -57,13 +55,32 @@ def load_port(name):
+
+
+# Called from test_sanity
+def validate_ports():
```
One reason I thought we might want to do this at runtime is that we might want to allow folks to drop in new ports (or maybe as you suggest using URL of a port on the command line).  In that case it might be good to validate each time you use a port.  However, we can address that once we get to that point.

ypujante replied:
Of course if/when ports are added dynamically or loaded on demand this would have to change.

sbc100 replied:
I mean technically you can do today by simply dropping a new file into `tools/ports`.. but I'm not sure anyone is doing that.

ypujante replied:
I reverted the change and now `validate_ports()` is called every time ports are read (line 99) for the potential cases where somebody might drop a file in `tools/ports`

ypujante commented:
@sbc100 I have made the changes you requested:

* make sure that `validate_ports` is called when ports are read to account for users potentially dropping a file under `tools/port`
* changed the name of the variable `name`.
 
But I want to get your feedback on this. Based of my understanding of the code, when calling `ports.fetch_project` you actually **MUST** call it with the port name (technically speaking, because there is NO enforcement you must call it with A valid port name, which obviously would make NO sense to use a different port name, but it is not enforced...). I did try to call it with something else, but it does fail, so this is a requirement.

And then you must use this same name in the rest of the code (like in `get_include_dir`). So I know you said that current port don't use `name` but in reality, you must provide the name of the port, so having the port name copy/pasted many times in the port code does not seem like good coding practice. For contrib ports, that means having the string `contrib.glfw3` copy/pasted 5 times which is why I would rather use a variable.

I named it `key` in the last iteration, but I am not sure it is accurate, because it must be the port name. Shall I call it `port_name` or simply use `name`?

sbc100 commented:
> @sbc100 I have made the changes you requested:
> 
> * make sure that `validate_ports` is called when ports are read to account for users potentially dropping a file under `tools/port`
> * changed the name of the variable `name`.
> 
> But I want to get your feedback on this. Based of my understanding of the code, when calling `ports.fetch_project` you actually **MUST** call it with the port name (technically speaking, because there is NO enforcement you must call it with A valid port name, which obviously would make NO sense to use a different port name, but it is not enforced...). I did try to call it with something else, but it does fail, so this is a requirement.
> 
> And then you must use this same name in the rest of the code (like in `get_include_dir`). So I know you said that current port don't use `name` but in reality, you must provide the name of the port, so having the port name copy/pasted many times in the port code does not seem like good coding practice. For contrib ports, that means having the string `contrib.glfw3` copy/pasted 5 times which is why I would rather use a variable.
> 
> I named it `key` in the last iteration, but I am not sure it is accurate, because it must be the port name. Shall I call it `port_name` or simply use `name`?

I would say just repeat yourself for this PR (like all the existing ports) and then we can consider ways to remove the duplication of the name string across all ports as a separate thing.

ypujante commented:
> I would say just repeat yourself for this PR (like all the existing ports) and then we can consider ways to remove the duplication of the name string across all ports as a separate thing.

I thought that contrib ports were different and handled by their authors... but fine if that it is what it takes...

sbc100 commented on test/test_other.py:
```diff
@@ -2369,6 +2369,11 @@ def test_sdl2_ttf(self):
+  def test_contrib_ports(self):
+    # Verify that contrib ports can be used (using the only contrib port available ATM, but can be replaced
+    # with a different contrib port when there is another one
+    self.emcc(test_file('other/test_contrib_ports.cpp'), ['--use-port=contrib.glfw3'], output_filename='a.out.js')
```
I think you can drop `output_filename='a.out.js'`

ypujante replied:
ok. I copy/pasted the million other tests around that do thit.

sbc100 replied:
Oh, sorry, actually this might be needed, sorry I didn't see this was a call to `self.emcc`.   I guess if the test passes without it that is fine.

ypujante replied:
it looks like it works without it on my local machine. I suppose we will find out if it fails on CI...

sbc100 commented on test/test_sanity.py:
```diff
@@ -527,9 +527,12 @@ def test_emconfig(self):
     restore_and_set_up()
 
+    # validating that ports are valid
+    ports.validate_ports()
```
I think you can remove this now that we do this on load.

ypujante replied:
Ok. Sure.

sbc100 commented on tools/link.py:
```diff
@@ -2963,8 +2965,6 @@ def add_js_deps(sym):
 
-  phase_calculate_system_libraries(state, linker_arguments, newargs)
-
   phase_link(linker_arguments, wasm_target, js_syms)
```
I think you can revert this file since this will be part of the "port settings" PR?

ypujante replied:
No this has nothing to do with port settings. It is a change that is required for using js_libraries in ports

sbc100 commented on tools/ports/contrib/glfw3.py:
```diff
@@ -0,0 +1,51 @@
+def linker_setup(ports, settings):
+  root_path = os.path.join(ports.get_dir(), 'contrib.glfw3')
+  source_js_path = os.path.join(root_path, 'src', 'js', 'lib_emscripten_glfw3.js')
+  settings.JS_LIBRARIES += [source_js_path]
```
This is never called as far as I can see?

ypujante replied:
Yes it is used. My port has a javascript implementation. If you remove this then link fails because it cannot find all the functions implemented in js

sbc100 commented on tools/ports/contrib/glfw3.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def process_args(ports):
+  return ['-isystem', ports.get_include_dir('contrib.glfw3')]
```
I wonder if we want the include path to include `contrib.` like this?    Maybe its fine and it avoid conflict the the builtin headers?  For ports that don't conflict I would hope we can avoid this.

Perhaps add a comment such as:

`Avoid installing headers into top level include path so that we don't conflict with the builtin GLFW headers that emscripten includes` 
?

ypujante replied:
Ok that makes sense

sbc100 commented on test/other/test_contrib_ports.cpp:
```diff
@@ -0,0 +1,25 @@
+ * found in the LICENSE file.
+ */
+
+#include <GLFW/glfw3.h>
```
Is this header any different the emscirpten one?

If not perhaps we could just install `emscripten_glfw3.h` and rely on the existing header?

ypujante replied:
I don't know if this it is the same or not. But the idea is that the port is self contained and brings its own dependencies. If it were to diverge then that would become an issue. I would rather keep it that way. Note that this `test_contrib_ports.cpp` will be replaced with another port once there is another one (when/if I work on libarchive I will switch it to this port)

sbc100 commented:
I wonder if we should perhaps pick a simpler port for our first one?  Something that doesn't conflict with an existing library and doesn't report JS library code?

ypujante commented:
> I wonder if we should perhaps pick a simpler port for our first one? Something that doesn't conflict with an existing library and doesn't report JS library code?

I made the changes you requested.

I have been spending a lot of time working on this and I am personally not interested in adding another port first. If your issue is because there is another glfw built-in port then it sounds like you will never accept my port in the first place. If you want me to rename it even further to make it even less confusing (like `contrib.glfw3_cpp`) I am happy to do that.

sbc100 reviewed (approved):
Fair enough, we can land this as is.

ypujante commented:
Thank you.

Like I said in my previous comment, I do have more in the pipe: options and libarchive (unless you were planning to work on that).

Other things that could be done: 
* loading (contrib) ports on demand only
* auto-generating contrib ports documentation

I would be happy to work on these but it doesn't look like it is an imminent need so just let me know if you ever want to work on these.

sbc100 commented:
> Thank you.
> 
> Like I said in my previous comment, I do have more in the pipe: options and libarchive (unless you were planning to work on that).

Sounds good!  Thanks for working on these things.

> 
> Other things that could be done:
> 
> * loading (contrib) ports on demand only
> * auto-generating contrib ports documentation
> 
> I would be happy to work on these but it doesn't look like it is an imminent need so just let me know if you ever want me to work on these.

I agree, neither of these seems urgent.

sbc100 commented:
BTW, I don't see any urgent need to add `libarchive`.   I just chose that because its very small simple C library without too much complexity.

ypujante commented:
> BTW, I don't see any urgent need to add `libarchive`. I just chose that because its very small simple C library without too much complexity.

Oh good because I was just looking at the project and it clearly does not look like a "C library without too much complexity" anymore! :)

ypujante commented:
On no, test failure :(

```
Out-of-date: npm packages
 -> /root/emsdk/node/16.20.0_64bit/bin/npm ci
npm ERR! code ECONNRESET
npm ERR! network aborted
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network 
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2024-02-05T21_13_04_962Z-debug-0.log
Traceback (most recent call last):
  File "./bootstrap.py", line 78, in <module>
    main(sys.argv[1:])
  File "./bootstrap.py", line 72, in main
    shared.run_process(cmd, cwd=utils.path_from_root())
  File "/root/project/tools/shared.py", line 130, in run_process
    ret = subprocess.run(cmd, check=check, input=input, *args, **kw)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/root/emsdk/node/16.20.0_64bit/bin/npm', 'ci']' returned non-zero exit status 1.
```

ypujante commented:
Looks like all the other tests have passed besides `test-wasm2js1` which seems to be related to a network issue and not the changes I have made.
